File: faucet/app/routes/faucet.py
"""
ZecKit Faucet - Funding Request Endpoint (REAL Transactions)
"""
from flask import Blueprint, jsonify, request, current_app
from datetime import datetime
import re

faucet_bp = Blueprint('faucet', __name__)


def validate_address(address: str) -> tuple:
    """Validate Zcash address format"""
    if not address:
        return False, "Address is required"
    
    # Transparent: t1 or t3 (mainnet), tm (testnet/regtest)
    # Shielded Sapling: zs1
    # Unified: u1
    
    if address.startswith('t'):
        if not re.match(r'^t[13m][a-zA-Z0-9]{33}$', address):
            return False, "Invalid transparent address format"
    elif address.startswith('zs1'):
        if len(address) < 78:
            return False, "Invalid sapling address format"
    elif address.startswith('u1'):
        if len(address) < 100:
            return False, "Invalid unified address format"
    else:
        return False, "Unsupported address type"
    
    return True, ""


@faucet_bp.route('/request', methods=['POST'])
def request_funds():
    """
    Request test funds from faucet - REAL BLOCKCHAIN TRANSACTION!
    """
    data = request.get_json()
    if not data:
        return jsonify({"error": "Invalid JSON"}), 400
    
    # Validate address
    to_address = data.get('address')
    is_valid, error_msg = validate_address(to_address)
    if not is_valid:
        return jsonify({"error": error_msg}), 400
    
    # Get amount
    try:
        amount = float(data.get('amount', current_app.config['FAUCET_AMOUNT_DEFAULT']))
        
        min_amount = current_app.config['FAUCET_AMOUNT_MIN']
        max_amount = current_app.config['FAUCET_AMOUNT_MAX']
        
        if amount < min_amount or amount > max_amount:
            return jsonify({
                "error": f"Amount must be between {min_amount} and {max_amount} ZEC"
            }), 400
    
    except (ValueError, TypeError):
        return jsonify({"error": "Invalid amount"}), 400
    
    # Check wallet ready
    wallet = current_app.faucet_wallet
    if not wallet:
        return jsonify({"error": "Faucet wallet not available"}), 503
    
    # Check balance
    balance = wallet.get_balance()
    if balance < amount:
        return jsonify({
            "error": f"Insufficient faucet balance (available: {balance} ZEC)"
        }), 503
    
    # Send REAL transaction
    try:
        result = wallet.send_to_address(
            to_address=to_address,
            amount=amount,
            memo=data.get('memo')
        )
        
        if not result.get("success"):
            return jsonify({
                "error": f"Transaction failed: {result.get('error')}"
            }), 500
        
        new_balance = wallet.get_balance()
        
        return jsonify({
            "success": True,
            "txid": result["txid"],
            "address": to_address,
            "amount": amount,
            "new_balance": float(new_balance),
            "timestamp": result["timestamp"],
            "message": f"Successfully sent {amount} ZEC. Verify TXID: {result['txid']}"
        }), 200
    
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@faucet_bp.route('/address', methods=['GET'])
def get_faucet_address():
    """Get the faucet's receiving address"""
    wallet = current_app.faucet_wallet
    
    if not wallet:
        return jsonify({"error": "Faucet wallet not available"}), 503
    
    return jsonify({
        "address": wallet.get_address("unified"),
        "balance": float(wallet.get_balance())
    }), 200


@faucet_bp.route('/sync', methods=['POST'])
def sync_wallet():
    """Manually trigger wallet sync"""
    wallet = current_app.faucet_wallet
    
    if not wallet:
        return jsonify({"error": "Faucet wallet not available"}), 503
    
    try:
        wallet.sync_wallet()
        return jsonify({
            "success": True,
            "message": "Wallet synced successfully",
            "current_balance": float(wallet.get_balance())
        }), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

File: zeckit-faucet/api/faucet.rs
use axum::{Json, extract::State};
use serde::{Deserialize, Serialize};
use serde_json::json;

use crate::AppState;
use crate::error::FaucetError;
use crate::validation::validate_address_via_zebra;

#[derive(Debug, Deserialize)]
pub struct FaucetRequest {
    address: String,
    amount: Option<f64>,
    memo: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct FaucetResponse {
    success: bool,
    txid: String,
    address: String,
    amount: f64,
    new_balance: f64,
    timestamp: String,
    network: String,
    message: String,
}

pub async fn request_funds(
    State(state): State<AppState>,
    Json(payload): Json<FaucetRequest>,
) -> Result<Json<FaucetResponse>, FaucetError> {
    // Validate address via Zebra RPC
    let validated_address = validate_address_via_zebra(
        &payload.address,
        &state.config.zebra_rpc_url,
    ).await?;

    // Get and validate amount
    let amount = payload.amount.unwrap_or(state.config.faucet_amount_default);
    
    if amount < state.config.faucet_amount_min || amount > state.config.faucet_amount_max {
        return Err(FaucetError::InvalidAmount(format!(
            "Amount must be between {} and {} ZEC",
            state.config.faucet_amount_min,
            state.config.faucet_amount_max
        )));
    }

    // Send transaction
    let mut wallet = state.wallet.write().await;
    let txid = wallet.send_transaction(&validated_address, amount, payload.memo).await?;

    // Get new balance
    let new_balance = wallet.get_balance().await?;

    Ok(Json(FaucetResponse {
        success: true,
        txid: txid.clone(),
        address: validated_address,
        amount,
        new_balance: new_balance.total_zec(),
        timestamp: chrono::Utc::now().to_rfc3339(),
        network: "regtest".to_string(),
        message: format!("Sent {} ZEC on regtest. TXID: {}", amount, txid),
    }))
}

pub async fn get_faucet_address(
    State(state): State<AppState>,
) -> Result<Json<serde_json::Value>, FaucetError> {
    let wallet = state.wallet.read().await;
    let address = wallet.get_unified_address().await?;
    let balance = wallet.get_balance().await?;

    Ok(Json(json!({
        "address": address,
        "balance": balance.total_zec(),
        "network": "regtest"
    })))
}

